content/blog/15-making-rust-attractive-for-writing-gtk-applications.md

```diff
@@ -0,0 +1,301 @@
+            obj.init_template();
+        }
+    }
+
+    impl ObjectImpl for SimpleWidget {
+        fn dispose(&self) {
+            // since gtk 4.8
+            self.obj().dispose_template(Self::Type);
+            // before gtk 4.8, for each direct child widget
+            // self.label.unparent();
+        }
+    }
+    impl WidgetImpl for SimpleWidget {}
+}
+glib::wrapper! {
+    pub struct SimpleWidget(ObjectSubclass<imp::SimpleWidget>)
+        @extends gtk::Widget;
+}
+
+#[gtk::template_callbacks]
+impl SimpleWidget {
+    #[template_callback]
+    fn on_clicked(&self, _button: &gtk::Button) {
+        println!("Clicked!");
+    }
+}
+```
+
+More details can be found in <https://gtk-rs.org/gtk4-rs/stable/latest/docs/gtk4_macros/derive.CompositeTemplate.html> & <https://gtk-rs.org/gtk4-rs/stable/latest/docs/gtk4_macros/attr.template_callbacks.html>.
+
+### Book, documentations, examples
+
+Providing only great bindings of GTK 4 sadly won't help newcomers trying to learn GTK for the first time despite the increasing number of apps written using `gtk4-rs`.
+
+Julian Hofer have been writing a "GUI development with Rust and GTK4" book that you can find at <https://gtk-rs.org/gtk4-rs/stable/latest/book/>.
+
+On top of that, we spent a good amount of time ensuring the documentation we provide, which is based on the C library documentation, has valid intra-links, uses the images provided by upstream to represent the various widgets and that most of the types/functions are properly documented.
+
+The `gtk4-rs` repository also includes ~35 examples that you can find in <https://github.com/gtk-rs/gtk4-rs/tree/master/examples> as well as a `GTK & Rust` template you can use to initialize your next project <https://gitlab.gnome.org/World/Rust/gtk-rust-template>.
+
+## Flatpak & Rust stable/nightly SDKs
+
+Beyond the building blocks provided by the bindings, we have also worked on providing a stable and a nightly Rust SDKs that can be used to either distribute your application as a Flatpak or during development time. The stable SDK comes with [Mold](https://github.com/rui314/mold) pre-installed as well, which is recommended if you want a slightly faster linking time.
+
+Flatpak can be used during development time with either [GNOME Builder](https://flathub.org/apps/details/org.gnome.Builder), VSCode with the [flatpak-vscode](https://marketplace.visualstudio.com/items?itemName=bilelmoussaoui.flatpak-vscode) extension or [fenv](https://gitlab.gnome.org/ZanderBrown/fenv/) if you prefer the CLI for building/running your application.
+
+- <https://github.com/flathub/org.freedesktop.Sdk.Extension.rust-stable>
+- <https://github.com/flathub/org.freedesktop.Sdk.Extension.rust-nightly>
+
+## Portals
+
+Modern Linux applications should make use of the [portals](https://flatpak.github.io/xdg-desktop-portal/) to request runtime permissions to access resources such as the camera, starting a screencast session or picking a file. 
```

Comment:
Well, arguably just using pipewire directly is equally fine, isn't it? :)

Reply:
For screencasting yes, i think so but from a sandboxed application you would need access to the pipewire socket which you don't have access to by default